### DISCUSSION TITLE
Git aliases

### DISCUSSION THREAD

**Discussion by mxz42** [fedimint/fedimint#1310] [General]
These might be some helpful git aliases to create fixup commits and squash theme automatically.

in your git config:
```
[alias]
    oldest-ancestor = !bash -c 'diff -u <(git rev-list --first-parent master) <(git rev-list --first-parent HEAD) | sed -ne \"s/^ //p\" | head -1'
    fix = "![ -z \"$1\" ] && echo Commit Hash required. || git commit --no-edit --fixup \"$1\" #"
    fixup = "!git rebase -i --autosquash $(git oldest-ancestor)"
```

- `oldest-ancestor` finds the commit where you forked your master branch from.
- `git fix <commit>` creates a fixup commit for the specified commit hash.
- `git fixup` then squashes your fixup commits into the proper commits. The "oldest-ancestor" is used for the rebase.

**Comment by dpc** [comment]
I do this with `git fetch upstream && git rebase -i upstream/master`.

**Comment by mxz42** [comment]
Right. I think the difference is that while you rebase on `upstream/master`, the `git fixup` uses the commit where the current branch was created from. Probably nice, if you just want to fix up some commits without rebasing on `upstream/master`.

**Comment by elsirion** [comment]
Establishing a fixup based workflow would be nice to reduce re-review. If someone joins review late they can probably even just apply the fixups themselves locally.

**Comment by dpc** [comment]
> Right. I think the difference is that while you rebase on upstream/master, the git fixup uses the commit where the current branch was created from. Probably nice, if you just want to fix up some commits without rebasing on upstream/master.


I don't understand the details there, but `git rebase -i master` or any other thing there works kind of too.

> Establishing a fixup based workflow would be nice to reduce re-review. If someone joins review late they can probably even just apply the fixups themselves locally.

I don't understand what you mean. All of this here is what people do locally, so it doesn't really help with reviewing PRs. Would be nice if github had some built-in workflow to "merge with squashing fixups" (is that what you meant?). I wouldn't not want to go through the hasle of doing it manually. Maybe bors can do such things...

**Comment by elsirion** [comment]
The workflow would be:
1. Push initial PR version
2. Receive review
3. Make changes as fixup commits, so they are easy to review for the original complainer
4. Once everyone is happy squash fixups into appropriate commits (can be automatically checked nothing else changed afaik)
5. Re-ACK and merge

If someone joins review after 3. they can also locally auto-squash all the fixups into their corresponding commits and view the PR that way (with already fixed commits).

The cool thing about 4. is that git can automatically squash fixups into their corresponding commits so you don't end up with one nasty fixup commit at the end but also don't have to interactively rebase all the time invalidating all previous review and making it unclear what changed.

**Comment by justinmoon** [comment]
That workflow sounds great.

**Comment by dpc** [comment]
I see we were trying this convention at my previous team, but there's no way to enforce it, or even guide people to use it, so it was a bit hit or miss. And then it has a cost of extra final step, which sometimes does not out-weights the benefits (change small enough for it not to matter). So I would encourage it, but I have no other ideas how to help implement it.
